File: web/src/features/batch/__tests__/medicationNameMatcher.test.ts
import { describe, expect, it } from 'vitest';
import {
  matchMedicationsInCommand,
  normalize,
  type Medication,
} from '../medicationNameMatcher';

const WELLBUTRIN: Medication = {
  id: '10',
  name: 'Wellbutrin XL',
  display_label: 'Bupropion',
};
const ADDERALL: Medication = { id: '20', name: 'Adderall' };
const TWO_WELLBUTRINS: Medication[] = [
  { id: '30', name: 'Wellbutrin' },
  { id: '31', name: 'Wellbutrin' },
];

describe('medicationNameMatcher (twin-checked vs Kotlin/Python)', () => {
  it('exact match returns unambiguous', () => {
    expect(
      matchMedicationsInCommand('I took my Wellbutrin XL today', [
        WELLBUTRIN,
        ADDERALL,
      ]),
    ).toEqual({ kind: 'unambiguous', matches: { 'wellbutrin xl': '10' } });
  });

  it('case mismatch normalizes', () => {
    expect(matchMedicationsInCommand('WELLBUTRIN xl', [WELLBUTRIN])).toEqual({
      kind: 'unambiguous',
      matches: { 'wellbutrin xl': '10' },
    });
  });

  it('trailing whitespace normalizes', () => {
    expect(matchMedicationsInCommand('  took adderall   ', [ADDERALL])).toEqual({
      kind: 'unambiguous',
      matches: { adderall: '20' },
    });
  });

  it('unicode smart quote normalizes', () => {
    expect(matchMedicationsInCommand('“took adderall”', [ADDERALL])).toEqual({
      kind: 'unambiguous',
      matches: { adderall: '20' },
    });
  });

  it('typo returns no match', () => {
    expect(matchMedicationsInCommand('took wellbutrn', [WELLBUTRIN])).toEqual({
      kind: 'no_match',
    });
  });

  it('two wellbutrins returns ambiguous', () => {
    const result = matchMedicationsInCommand('took my Wellbutrin', TWO_WELLBUTRINS);
    expect(result).toEqual({
      kind: 'ambiguous',
      phrases: [{ phrase: 'wellbutrin', candidate_entity_ids: ['30', '31'] }],
    });
  });

  it('display label match when name does not', () => {
    expect(
      matchMedicationsInCommand('finished bupropion today', [WELLBUTRIN]),
    ).toEqual({ kind: 'unambiguous', matches: { bupropion: '10' } });
  });

  it('mixed command with one unambiguous and one ambiguous returns mixed', () => {
    const result = matchMedicationsInCommand(
      'took my Wellbutrin and Adderall',
      [ADDERALL, ...TWO_WELLBUTRINS],
    );
    expect(result).toEqual({
      kind: 'mixed',
      unambiguous: { adderall: '20' },
      ambiguous: [{ phrase: 'wellbutrin', candidate_entity_ids: ['30', '31'] }],
    });
  });

  it('trailing punctuation is stripped', () => {
    expect(matchMedicationsInCommand('took my Adderall.', [ADDERALL])).toEqual({
      kind: 'unambiguous',
      matches: { adderall: '20' },
    });
  });

  it('longest key wins on overlap', () => {
    const plain: Medication = { id: '100', name: 'Wellbutrin' };
    const xl: Medication = { id: '101', name: 'Wellbutrin XL' };
    expect(matchMedicationsInCommand('took my Wellbutrin XL', [plain, xl])).toEqual({
      kind: 'unambiguous',
      matches: { 'wellbutrin xl': '101' },
    });
  });

  it('substring inside another word does not match', () => {
    expect(matchMedicationsInCommand('took my addy', [ADDERALL])).toEqual({
      kind: 'no_match',
    });
  });

  it('empty command returns no match', () => {
    expect(matchMedicationsInCommand('', [ADDERALL])).toEqual({ kind: 'no_match' });
    expect(matchMedicationsInCommand('   ', [ADDERALL])).toEqual({ kind: 'no_match' });
  });

  it('empty medication list returns no match', () => {
    expect(matchMedicationsInCommand('took adderall', [])).toEqual({ kind: 'no_match' });
  });

  it('normalize is idempotent', () => {
    const once = normalize('  Wellbutrin XL!  ');
    const twice = normalize(once);
    expect(once).toBe('wellbutrin xl');
    expect(twice).toBe(once);
  });
});

File: app/src/test/java/com/averycorp/prismtask/domain/usecase/MedicationNameMatcherTest.kt
package com.averycorp.prismtask.domain.usecase

import com.averycorp.prismtask.domain.usecase.MedicationNameMatcher.AmbiguousPhrase
import com.averycorp.prismtask.domain.usecase.MedicationNameMatcher.MatchResult
import com.averycorp.prismtask.domain.usecase.MedicationNameMatcher.Medication
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Twin-checked against the Python and TS matchers — the same input must
 * produce the same `MatchResult` shape on every surface.
 */
class MedicationNameMatcherTest {

    private val wellbutrin = Medication(id = "10", name = "Wellbutrin XL", displayLabel = "Bupropion")
    private val adderall = Medication(id = "20", name = "Adderall")
    private val twoWellbutrins = listOf(
        Medication(id = "30", name = "Wellbutrin"),
        Medication(id = "31", name = "Wellbutrin")
    )

    @Test
    fun `exact match returns unambiguous`() {
        val result = MedicationNameMatcher.match("I took my Wellbutrin XL today", listOf(wellbutrin, adderall))
        assertEquals(MatchResult.Unambiguous(mapOf("wellbutrin xl" to "10")), result)
    }

    @Test
    fun `case mismatch normalizes`() {
        val result = MedicationNameMatcher.match("WELLBUTRIN xl", listOf(wellbutrin))
        assertEquals(MatchResult.Unambiguous(mapOf("wellbutrin xl" to "10")), result)
    }

    @Test
    fun `trailing whitespace normalizes`() {
        val result = MedicationNameMatcher.match("  took adderall   ", listOf(adderall))
        assertEquals(MatchResult.Unambiguous(mapOf("adderall" to "20")), result)
    }

    @Test
    fun `unicode smart quote normalizes`() {
        // The smart-quote stays as-is after NFC; the matcher normalizes case
        // and trailing punctuation only — quotes inside the command don't
        // break a whole-word match because they're non-word boundaries.
        val result = MedicationNameMatcher.match("“took adderall”", listOf(adderall))
        assertEquals(MatchResult.Unambiguous(mapOf("adderall" to "20")), result)
    }

    @Test
    fun `typo returns no match`() {
        // "wellbutrn" is one letter off — exact contract refuses to guess.
        val result = MedicationNameMatcher.match("took wellbutrn", listOf(wellbutrin))
        assertEquals(MatchResult.NoMatch, result)
    }

    @Test
    fun `two wellbutrins returns ambiguous`() {
        val result = MedicationNameMatcher.match("took my Wellbutrin", twoWellbutrins)
        assertTrue(result is MatchResult.Ambiguous)
        result as MatchResult.Ambiguous
        assertEquals(1, result.phrases.size)
        assertEquals(AmbiguousPhrase("wellbutrin", listOf("30", "31")), result.phrases[0])
    }

    @Test
    fun `display label match when name does not`() {
        val result = MedicationNameMatcher.match("finished bupropion today", listOf(wellbutrin))
        assertEquals(MatchResult.Unambiguous(mapOf("bupropion" to "10")), result)
    }

    @Test
    fun `mixed command with one unambiguous and one ambiguous returns mixed`() {
        val meds = listOf(adderall) + twoWellbutrins
        val result = MedicationNameMatcher.match("took my Wellbutrin and Adderall", meds)
        assertTrue(result is MatchResult.Mixed)
        result as MatchResult.Mixed
        assertEquals(mapOf("adderall" to "20"), result.unambiguous)
        assertEquals(1, result.ambiguous.size)
        assertEquals(AmbiguousPhrase("wellbutrin", listOf("30", "31")), result.ambiguous[0])
    }

    @Test
    fun `trailing punctuation is stripped`() {
        val result = MedicationNameMatcher.match("took my Adderall.", listOf(adderall))
        assertEquals(MatchResult.Unambiguous(mapOf("adderall" to "20")), result)
    }

    @Test
    fun `longest key wins on overlap`() {
        val plain = Medication(id = "100", name = "Wellbutrin")
        val xl = Medication(id = "101", name = "Wellbutrin XL")
        val result = MedicationNameMatcher.match("took my Wellbutrin XL", listOf(plain, xl))
        assertEquals(MatchResult.Unambiguous(mapOf("wellbutrin xl" to "101")), result)
    }

    @Test
    fun `substring inside another word does not match`() {
        // "addy" is a slang for Adderall, but the matcher must not infer it
        // — substring/fuzzy match is failure mode #2, which is the bug.
        val result = MedicationNameMatcher.match("took my addy", listOf(adderall))
        assertEquals(MatchResult.NoMatch, result)
    }

    @Test
    fun `empty command returns no match`() {
        assertEquals(MatchResult.NoMatch, MedicationNameMatcher.match("", listOf(adderall)))
        assertEquals(MatchResult.NoMatch, MedicationNameMatcher.match("   ", listOf(adderall)))
    }

    @Test
    fun `empty medication list returns no match`() {
        assertEquals(MatchResult.NoMatch, MedicationNameMatcher.match("took adderall", emptyList()))
    }
}
